File: C++/cyclic_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>

class CyclicSort {
private:
    std::vector<int> arr;

    void swap(int& a, int& b) {
        int temp = a;
        a = b;
        b = temp;
    }

public:
    // Constructor to initialize with random values from 1 to size
    CyclicSort(size_t size) {
        // Use modern random number generation
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(1, size);

        arr.reserve(size);
        for(size_t i = 0; i < size; ++i) {
            arr.push_back(dis(gen));
        }
    }

    // Cyclic sort implementation
    void sort() {
        int i = 0;
        while (i < arr.size()) {
            // Since array contains numbers from 1 to N
            // correct position for element arr[i] is arr[i] - 1
            int correct_pos = arr[i] - 1;
            
            if (arr[i] != arr[correct_pos]) {
                swap(arr[i], arr[correct_pos]);
            } else {
                i++;
            }
        }
    }

    // Display array content
    void displayArray(const std::string& message = "") const {
        std::cout << message;
        for (const auto& num : arr) {
            std::cout << num << " ";
        }
        std::cout << "\n";
    }

    // Verify if array is sorted
    bool isSorted() const {
        for (size_t i = 1; i < arr.size(); ++i) {
            if (arr[i] < arr[i-1]) return false;
        }
        return true;
    }
};

int main() {
    const size_t ARRAY_SIZE = 10;
    CyclicSort sorter(ARRAY_SIZE);
    
    // Display original array
    sorter.displayArray("Original array: ");
    
    // Perform cyclic sort
    sorter.sort();
    
    // Display sorted array
    sorter.displayArray("Sorted array:   ");
    
    // Verify sorting
    std::cout << "\nArray is " << (sorter.isSorted() ? "correctly" : "not") 
              << " sorted!" << std::endl;

    return 0;
}

File: Java/cyclicsort.java
import java.util.Arrays;

public class CycleSort {

    // Function to perform Cycle Sort
    public static void cycleSort(int[] arr) {
        int n = arr.length;

        // Traverse the array to place each element at its correct position
        for (int cycleStart = 0; cycleStart <= n - 2; cycleStart++) {
            int item = arr[cycleStart];

            // Find position where we put the item
            int pos = cycleStart;
            for (int i = cycleStart + 1; i < n; i++) {
                if (arr[i] < item) {
                    pos++;
                }
            }

            // If the item is already in correct position
            if (pos == cycleStart) {
                continue;
            }

            // Skip duplicates
            while (item == arr[pos]) {
                pos++;
            }

            // Swap item with correct position
            if (pos != cycleStart) {
                int temp = item;
                item = arr[pos];
                arr[pos] = temp;
            }

            // Rotate the rest of the cycle
            while (pos != cycleStart) {
                pos = cycleStart;
                for (int i = cycleStart + 1; i < n; i++) {
                    if (arr[i] < item) {
                        pos++;
                    }
                }

                while (item == arr[pos]) {
                    pos++;
                }

                if (item != arr[pos]) {
                    int temp = item;
                    item = arr[pos];
                    arr[pos] = temp;
                }
            }
        }
    }

    public static void main(String[] args) {
        int[] arr = {20, 40, 50, 10, 30};
        System.out.println("Original array: " + Arrays.toString(arr));

        cycleSort(arr);

        System.out.println("Sorted array:   " + Arrays.toString(arr));
    }
}
